bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/internal/OpenThermGatewaySocketConnector.java

```diff
@@ -0,0 +1,219 @@
+                // Make writer accessible on class level
+                writer = wrt;
+
+                sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintReport, "A"));
+                // Set the OTGW to report every message it receives and transmits
+                sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintSummary, "0"));
+
+                while (!stopping && !Thread.currentThread().isInterrupted()) {
+                    @Nullable
+                    String message = reader.readLine();
+
+                    if (message != null) {
+                        handleMessage(message);
+                    } else {
+                        logger.debug("Connection closed by OpenTherm Gateway");
+                        break;
+                    }
+                }
+
+                logger.debug("Stopping OpenThermGatewaySocketConnector");
+            } finally {
+                connected = false;
+
+                logger.debug("OpenThermGatewaySocketConnector disconnected");
+                callback.disconnected();
+            }
+        } catch (IOException ex) {
+            logger.warn("Unable to connect to the OpenTherm Gateway.", ex);
+        }
+    }
+
+    @Override
+    public void stop() {
+        logger.debug("Stopping OpenThermGatewaySocketConnector");
+        stopping = true;
+    }
+
+    @Override
+    public boolean isConnected() {
+        return connected;
+    }
+
+    @Override
+    public void sendCommand(GatewayCommand command) {
+        @Nullable
+        PrintWriter wrtr = writer;
+
+        String msg = command.toFullString();
+
+        pendingCommands.put(command.getCode(),
+                new AbstractMap.SimpleImmutableEntry<>(System.currentTimeMillis(), command));
+
+        if (connected) {
+            logger.debug("Sending message: {}", msg);
+            if (wrtr != null) {
+                wrtr.print(msg + "\r\n");
+                wrtr.flush();
+            }
+        } else {
+            logger.debug("Unable to send message: {}. OpenThermGatewaySocketConnector is not connected.", msg);
+        }
+    }
+
+    private void handleMessage(String message) {
+        if (message.length() > 2 && message.charAt(2) == ':') {
+            String code = message.substring(0, 2);
+            String value = message.substring(3);
+
+            logger.debug("Received command confirmation: {}: {}", code, value);
+            pendingCommands.remove(code);
+            return;
+        }
+
+        long currentTime = System.currentTimeMillis();
+
+        for (Entry<Long, GatewayCommand> timeAndCommand : pendingCommands.values()) {
+            long responseTime = timeAndCommand.getKey() + COMMAND_RESPONSE_TIME_MILLISECONDS;
+            long timeoutTime = timeAndCommand.getKey() + COMMAND_TIMEOUT_MILLISECONDS;
+
+            if (currentTime > responseTime && currentTime <= timeoutTime) {
+                logger.debug("Resending command: {}", timeAndCommand.getValue().toFullString());
```

Comment:
You don't have to call `toString` it will be called automatically, but only when the logging-level actually applies.

Reply:
I am not calling toString(), but toFullString() which takes care of formatting.

Reply:
Yes, but now that method will also be called while debugging is not enabled, the question is whether that is desirable. Leaving it out might make the formatting a little bit more ugly but it will not execute formatting of a String intended for debugging while the system is not being executed with debugging enabled.

Reply:
removed toFullString()